File: kotlin/openid4vp/src/testFixtures/kotlin/com/hopae/eudi/wallet/vp/MockMdocVerifier.kt
package com.hopae.eudi.wallet.vp

import com.hopae.eudi.wallet.cbor.Cbor
import com.hopae.eudi.wallet.cbor.CborDecoder
import com.hopae.eudi.wallet.cbor.CborEncoder
import com.hopae.eudi.wallet.cbor.cose.CoseSign1
import com.hopae.eudi.wallet.mdoc.IssuerSigned
import com.hopae.eudi.wallet.sdjwt.Base64Url
import com.hopae.eudi.wallet.sdjwt.JsonValue
import com.hopae.eudi.wallet.spi.HttpMethod
import com.hopae.eudi.wallet.spi.HttpRequest
import com.hopae.eudi.wallet.spi.HttpResponse
import com.hopae.eudi.wallet.spi.HttpTransport
import java.net.URLDecoder
import java.net.URLEncoder

/**
 * A mock OpenID4VP verifier for mdoc: serves an mso_mdoc DCQL query and verifies the returned
 * DeviceResponse — device signature over the OID4VP SessionTranscript, using the deviceKey from the MSO.
 * Unencrypted (`direct_post`) for simplicity.
 */
class MockMdocVerifier(
    val docType: String = "org.iso.18013.5.1.mDL",
    val namespace: String = "org.iso.18013.5.1",
) : HttpTransport {
    val clientId = "verifier.example"
    val nonce = "vp-nonce-mdoc"
    val responseUri = "https://verifier.example/response"

    /** Element identifiers disclosed by the last verified DeviceResponse. */
    var disclosedElements: Set<String>? = null
        private set

    fun requestUri(): String {
        val dcql = """{"credentials":[{"id":"mdl","format":"mso_mdoc","meta":{"doctype_value":"$docType"},""" +
            """"claims":[{"path":["$namespace","family_name"]},{"path":["$namespace","given_name"]}]}]}"""
        return "openid4vp://?client_id=${enc(clientId)}&nonce=${enc(nonce)}" +
            "&response_mode=direct_post&response_uri=${enc(responseUri)}&state=xyz&dcql_query=${enc(dcql)}"
    }

    override suspend fun execute(request: HttpRequest): HttpResponse {
        if (request.url != responseUri || request.method != HttpMethod.POST) return HttpResponse(404, emptyList(), ByteArray(0))
        val form = request.body!!.decodeToString().split('&').associate {
            URLDecoder.decode(it.substringBefore('='), "UTF-8") to URLDecoder.decode(it.substringAfter('='), "UTF-8")
        }
        val vpToken = JsonValue.parse(form["vp_token"]!!) as JsonValue.Obj
        val presentation = ((vpToken["mdl"] as JsonValue.Arr).items.first() as JsonValue.Str).value
        val deviceResponse = CborDecoder.decode(Base64Url.decode(presentation))

        val document = (map(deviceResponse, "documents") as Cbor.Array).items.single()
        val issuerSigned = IssuerSigned.fromCbor(map(document, "issuerSigned")!!)
        val deviceKey = issuerSigned.parseMso().deviceKey

        // device signature over the reconstructed DeviceAuthenticationBytes (thumbprint null for direct_post)
        val deviceSigned = map(document, "deviceSigned")!!
        val deviceSignature = CoseSign1.fromCbor(map(map(deviceSigned, "deviceAuth")!!, "deviceSignature")!!)
        val sessionTranscript = Oid4vpSessionTranscript.build(clientId, responseUri, nonce, null)
        val deviceNameSpacesBytes = map(deviceSigned, "nameSpaces")!!
        val deviceAuth = Cbor.Array(listOf(Cbor.Text("DeviceAuthentication"), sessionTranscript, Cbor.Text(docType), deviceNameSpacesBytes))
        val deviceAuthBytes = CborEncoder.encode(Cbor.Tagged(24u, Cbor.Bytes(CborEncoder.encode(deviceAuth))))
        require(deviceSignature.verify(deviceKey, detachedPayload = deviceAuthBytes)) { "device signature must verify" }

        disclosedElements = issuerSigned.nameSpaces[namespace]!!.map { it.item.elementIdentifier }.toSet()
        return HttpResponse(200, listOf("Content-Type" to "application/json"), """{"redirect_uri":"https://verifier.example/done"}""".encodeToByteArray())
    }

    private fun map(c: Cbor, key: String): Cbor? =
        (c as Cbor.CborMap).entries.firstOrNull { (k, _) -> (k as? Cbor.Text)?.value == key }?.second

    private companion object {
        fun enc(s: String) = URLEncoder.encode(s, "UTF-8")
    }
}

File: swift/Sources/WalletTestKit/MockMdocVerifier.swift
import CborCose
import Foundation
import MDoc
import OpenID4VP
import SdJwt
import WalletAPI

/// A mock OpenID4VP verifier for mdoc: serves an mso_mdoc DCQL query and verifies the returned
/// DeviceResponse — device signature over the OID4VP SessionTranscript, using the deviceKey from the MSO.
/// Unencrypted (`direct_post`).
public actor MockMdocVerifier: HttpTransport {
    public let docType: String
    public let namespace: String
    public let clientId = "verifier.example"
    public let nonce = "vp-nonce-mdoc"
    public let responseUri = "https://verifier.example/response"
    public private(set) var disclosedElements: Set<String>?

    public init(docType: String = "org.iso.18013.5.1.mDL", namespace: String = "org.iso.18013.5.1") {
        self.docType = docType
        self.namespace = namespace
    }

    public func requestUri() -> String {
        let dcql = "{\"credentials\":[{\"id\":\"mdl\",\"format\":\"mso_mdoc\",\"meta\":{\"doctype_value\":\"\(docType)\"}," +
            "\"claims\":[{\"path\":[\"\(namespace)\",\"family_name\"]},{\"path\":[\"\(namespace)\",\"given_name\"]}]}]}"
        return "openid4vp://?client_id=\(enc(clientId))&nonce=\(enc(nonce))&response_mode=direct_post" +
            "&response_uri=\(enc(responseUri))&state=xyz&dcql_query=\(enc(dcql))"
    }

    public func execute(_ request: HttpRequest) async throws -> HttpResponse {
        guard request.url == responseUri, request.method == .post else {
            return HttpResponse(status: 404, headers: [], body: [])
        }
        let bodyStr = String(bytes: request.body ?? [], encoding: .utf8) ?? ""
        var form: [String: String] = [:]
        for pair in bodyStr.split(separator: "&") {
            let kv = pair.split(separator: "=", maxSplits: 1)
            form[String(kv[0]).removingPercentEncoding ?? ""] = kv.count > 1 ? (String(kv[1]).removingPercentEncoding ?? "") : ""
        }
        let vpToken = try JsonValue.parse(form["vp_token"]!)
        guard case let .arr(items)? = vpToken["mdl"], case let .str(presentation) = items[0] else {
            throw VpError.responseFailed("no mdl presentation")
        }
        let deviceResponse = try CborDecoder.decode(Base64Url.decode(presentation))
        guard case let .array(documents)? = field(deviceResponse, "documents") else {
            throw VpError.responseFailed("no documents")
        }
        let document = documents[0]
        let issuerSigned = try IssuerSigned.fromCbor(field(document, "issuerSigned")!)
        let deviceKey = try issuerSigned.parseMso().deviceKey

        // device signature over the reconstructed DeviceAuthenticationBytes (thumbprint nil for direct_post)
        let deviceSigned = field(document, "deviceSigned")!
        let deviceSignature = try CoseSign1.fromCbor(field(field(deviceSigned, "deviceAuth")!, "deviceSignature")!)
        let st = try Oid4vpSessionTranscript.build(clientId: clientId, responseUri: responseUri, nonce: nonce, verifierJwkThumbprint: nil)
        let deviceNameSpacesBytes = field(deviceSigned, "nameSpaces")!
        let deviceAuth = Cbor.array([.text("DeviceAuthentication"), st, .text(docType), deviceNameSpacesBytes])
        let deviceAuthBytes = try CborEncoder.encode(.tagged(24, .bytes(try CborEncoder.encode(deviceAuth))))
        guard deviceSignature.verify(publicKey: deviceKey, detachedPayload: deviceAuthBytes) else {
            throw VpError.responseFailed("device signature failed")
        }

        disclosedElements = Set(issuerSigned.nameSpaces.first(where: { $0.0 == namespace })?.1.map { $0.item.elementIdentifier } ?? [])
        return HttpResponse(status: 200, headers: [("Content-Type", "application/json")],
                            body: [UInt8](#"{"redirect_uri":"https://verifier.example/done"}"#.utf8))
    }

    private func field(_ c: Cbor, _ key: String) -> Cbor? {
        guard case let .map(entries) = c else { return nil }
        return entries.first(where: { if case let .text(k) = $0.0 { return k == key }; return false })?.1
    }

    private nonisolated func enc(_ s: String) -> String {
        s.addingPercentEncoding(withAllowedCharacters: .alphanumerics) ?? s
    }
}
